### PR TITLE
Automatically assign roles from #request-rank

### PR DESCRIPTION
**Pull request by CaptainDestructo**
Smaller items first:

- Moved `all_lpd_ranks` definition to the loop in Officer Manager, that way if for some reason the hierarchy is updated in Discord, it will be updated in the bot on the next loop. --- Added `Tuple(bot.officer_manager.all_lpd_ranks)` to available variables.
- Added `Offier.rank` property, which returns the Officer's `Discord.Role` object having the highest position in the role hierarchy that also appears in `self.bot.officer_manager.all_lpd_ranks`.

Now for the main event:

When a message is sent in the `#request-rank` channel as specified in the settings file, process the message.

Messages will be processed under the following conditions:

- Message must be sent by an LPD cadet or officer (verified by presence of Officer object)
- Message must contain at least one mention
- Message must contain case-insensitive substrings `'trained by'` and `'request rank'`

If processing proceeds, the bot will react to the message with a ✅ for the one of the mentioned trainers to click within the specified timeout (default 300 seconds).

The bot will iterate through an internally defined dictionary of requestable roles, and select the first one where the role name appears in the message. This will cause unknown behavior if multiple roles are specified. Theoretically, it would process as if the first role in the internal dictionary were found.

During processing, the bot will check the position of the user's current LPD rank role against the position of the `prereq` role in the dictionary, as well as the `upgrade` boolean to determine whether or not the user has the correct rank to request that particular role. 

> If the test fails, the user's message will be deleted, they will be notified that they have insufficient rank, and the function will return without assigning or removing roles.

The bot will then gather a list of Officer IDs which are allowed to approve the request. This list contains the IDs of mentioned users having the specified `approver` role. When one of those officers clicks the ✅ within the timeout period, the bot will assign the user the requested role, and then delete the `prereq` role if `upgrade` is True.

> After an allowed approver clicks the ✅ and roles are assigned/removed, or the timeout period expires, whichever comes first, the function will return.

If the function has not returned after iterating through the dictionary, it means that a non-standard rank was requested. In this case, the bot will react with a ❓, and the function will return.

Other processing functions in this bot would usually delete the message and notify the user that there is a problem, however in this case I have opted to include forward compatibility for future ranks. If ranks are added, Staff will still be able to manually approve requests until the rank is added to the internal dictionary.

### CODE REVIEW

**Review by hrolfurgylfa — CHANGES_REQUESTED** [review]
Except for these few things, the code looked really good, and these were really minor anyway. I haven't gotten the time to try running it yet though, I will go ahead and try that sometime tomorrow.

File: Classes/extra_functions.py [hunk 1]
```diff
@@ -155,6 +155,163 @@ async def clean_shutdown(bot, location="the console", person="KeyboardInterrupt"
...
+            user.id in valid_approvers
+            and reaction.emoji == "\N{WHITE HEAVY CHECK MARK}"
+        ):
+            valid_approvers.remove(user.id)
+            return reaction, user
+
+    for key in requestables.keys():
+        if key in message.content.lower():
+
+            # current_rank = list(set(message.author.roles) & set(bot.officer_manager.all_lpd_ranks))[0]
```
[the remark this answers is not in the document]
Comment (hrolfurgylfa) [reply Classes/extra_functions.py:280]: Sorry, this was only supposed to reference line 280

[the remark this answers is not in the document]
Comment (CaptainDestructo) [reply Classes/extra_functions.py:280]: Removed line 280

File: Classes/extra_functions.py [hunk 2]
```diff
@@ -155,6 +155,163 @@ async def clean_shutdown(bot, location="the console", person="KeyboardInterrupt"
...
+
+    lmt_trained_role = bot.officer_manager.guild.get_role(
+        bot.settings["lmt_trained_role"]
+    )
+    slrt_trained_role = bot.officer_manager.guild.get_role(
+        bot.settings["slrt_trained_role"]
+    )
+    watch_officer_role = bot.officer_manager.guild.get_role(
+        bot.settings["watch_officer_role"]
+    )
```
[the remark this answers is not in the document]
Comment (CaptainDestructo) [reply Classes/extra_functions.py]: I wasn't aware black had this feature. I disabled formatting for this section.

File: Classes/commands.py [hunk 3]
```diff
@@ -1275,6 +1275,46 @@ async def strike(self, ctx):
...
+    @checks.is_white_shirt()
+    @commands.command()
+    async def promote(self, ctx):
+        """Promote the mentioned officers"""
+
+        for member in ctx.message.mentions:
+            officer = self.bot.officer_manager.get_officer(member.id)
+            try:
+                updated_rank = await officer.promote()
+                pattern = re.compile(r"(LPD )?(\w+( \w+)*)")
```
[the remark this answers is not in the document]
Comment (CaptainDestructo) [reply Classes/commands.py]: Moved to top of command outside loop

**Review by hrolfurgylfa — APPROVED** [review]
Seems good to me

**Pull request merged**